engine/src/main/java/org/terasology/internet/internal/InternetManagerImpl.java

```diff
@@ -26,32 +35,63 @@
 import java.security.AccessController;
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
-import java.util.Set;
-import java.util.TreeSet;
+import java.util.Collections;
+import java.util.Map;
 
 /**
  * Implements InternetManager.
  */
 public class InternetManagerImpl implements InternetManager {
+    /**
+     * The module this InternetManager belongs to.
+     */
+    //@Nonnull
     private final Name moduleId;
 
-    private final Set<String> allowedHosts;
+    /**
+     * The allowed hosts.
+     */
+    //@Nonnull
+    private final Map<String, TIntSet> allowedHosts;
 
     /**
      * Construct a new InternetManagerImpl with a shared "permanent" cache.
      *
      * @param moduleId The module ID.
+     * @param context The current context.
      */
-    public InternetManagerImpl(Name moduleId) {
+    public InternetManagerImpl(Name moduleId, Context context) {
         this.moduleId = moduleId;
-        allowedHosts = new TreeSet<>();
-        allowedHosts.add("irc.freenode.net");
+        Config theConfig = context.get(Config.class);
+        if (theConfig == null) {
+            throw new IllegalStateException();
+        }
+        // TODO check if this could miss config updates, and make thread safe.
+        InternetConfig.Hosts hosts = theConfig.getInternet().getTcpWhitelists().get(moduleId);
+        if (hosts == null) {
+            allowedHosts = Collections.emptyMap();
+            return;
+        }
+        allowedHosts = Maps.transformValues(hosts.getHosts(), (x) -> x != null ? TCollections.unmodifiableSet(x.getPorts()) : null);
     }
 
     @Override
     public TCPSocket openTCPConnection(String hostname, int port) throws IOException {
-        if (!allowedHosts.contains(hostname)) {
-            throw new IllegalArgumentException();
+        Map<String, TIntSet> allowedHosts = this.allowedHosts;
+        if (!(allowedHosts.containsKey(hostname) || allowedHosts.containsKey("*"))) {
+            // neither hostname nor wildcard in allowedHosts
+            // TODO maybe use a different exception type? (SandboxException or something?)
```

Comment:
Sounds good - `IllegalArgumentException` is IMO best used in `Preconditions`-esque checks

```java
public int foo (int a) {
    if (a % 2 == 2) {
        throw new IllegalArgumentException ("wow you messed up");
    }
    /* ... */
}
```

and I can't find a good default exception for this usecase.